Make inventory UI great again

stilnat opened the pull request:
<!-- The notes within these arrows are for you but can be deleted. -->

## Summary

Adds a functionning UI for the hands, and some placeholder for the ID, Bags, and pockets.
Reintroduce the sprite generator.
Make a few small cosmetic changes on the Inventory UI (rotation, outliner, spacing, size...).

## Pictures/Videos (optional)

![image](https://user-images.githubusercontent.com/14344825/208305761-ef805c83-fddb-481c-90a1-8b57592f39e2.png)


## Technical Notes (optional)

No major changes. The main change is that Inventory is now activating inventoryUI.

## Known issues

- The hands can't be switched by clicking on the icons like on the outdated branch. This can be solved in a different PR.
- The sprite are kind of ugly, could be due to the shader, maybe that can also be part of another PR.


## Fixes (optional)

<!-- List any issues or other PRs connected to this one. -->

<!-- If this PR CLOSES any issues/PRs, add "Closes" before the number (e.g. "Closes #123"). -->

Closes #935 
Closes #939

stilnat committed:
Adds back the ContainerDisplay prefab

This prefab is necessary to display the content of containers.

stilnat committed:
Adds back RunTimePreviewGenerator

Necessary to generate the sprites for items in containers.

stilnat committed:
Change the Generic panel prefab appearance

Small cosmetic changes

stilnat committed:
Make prefab for hands UI

stilnat committed:
Update the Humanoid Inventory

stilnat committed:
Activate InventoryUI on Inventory awakening

The InventoryUI script needs the Inventory script to be awaken to function.
It makes sense to let the inventory script handle the active state of the inventory UI script.
Also adds a reference toward the Inventory UI script, because the inventory should know what is it's UI.

stilnat committed:
Adds a reference toward the handsUI script in the Hand script

The hand script should know what is it's UI.

stilnat committed:
Put back item sprite generation

stilnat committed:
Merge branch 'RE-SS3D:develop' into MakeInventoryUIGreatAgain

cosmiccoincidence commented:
> bring an item from one hand over to the other and that item's ui pretty much breaks and you can't drag anything anymore. - LightVelox

Seems there seems to be some bugs. Is that due to lack of container functionality or a problem with the UI itself?

stilnat committed:
Merge branch 'RE-SS3D:develop' into MakeInventoryUIGreatAgain

stilnat commented:
> > bring an item from one hand over to the other and that item's ui pretty much breaks and you can't drag anything anymore. - LightVelox
> 
> Seems there seems to be some bugs. Is that due to lack of container functionality or a problem with the UI itself?

I'll check that as soon as I can get in game again, I have a bunch of assembly errors right now

stilnat commented:
@cosmiccoincidence Where did you see LightVelox comment ? I would have missed it if you didn't cite it

Ryan089 commented:
@stilnat it was in the #pull-requests channel [here](https://discord.com/channels/483813139633143808/483814334770839573/1058220429321457766).

TaylorNAlbarnaz committed:
Fix dragging and makes var names clearer

stilnat committed:
Fix falling items when swapping hands (or any containers)

TaylorNAlbarnaz reviewed (changes requested):
You should only be able to drag items with your Left Mouse Button, currently you can drag using the Right Mouse Button as well

stilnat committed:
Only allow to drag with a left click

stilnat commented:
Don't accept this PR yet, it's not working client side, only host side

stilnat commented:
Nevermind, the bug is only in the client when client is in editor

stilnat committed:
Merge branch 'RE-SS3D:develop' into MakeInventoryUIGreatAgain

stilnat committed:
Right dragging a sprite on slot doesn't move item anymore.

joaoburatto commented on Assets/Scripts/SS3D/Systems/Storage/Containers/AttachedContainer.cs:
```diff
@@ -150,7 +150,9 @@ void handleItemAdded(Item item)
             {
-                item.Unfreeze();
+                // Only unfreeze the item if it was not just placed into another container
+                if(item.Container == null)
+                    item.Unfreeze();
```
it is preferred to add braces

joaoburatto commented on Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs:
```diff
@@ -45,9 +47,13 @@ public class Hands : InteractionSourceNetworkBehaviour, IToolHolder, IInteractio
         /// </summary>
         public bool SelectedHandEmpty => SelectedHandContainer.Empty;
 
+        public HandsUi HandsUI {get; private set;}
```
add space before get and after set;

joaoburatto commented on Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs:
```diff
@@ -251,6 +252,22 @@ public void SetContainerUnchecked(Container newContainer)
+        {
+            RuntimePreviewGenerator.BackgroundColor = new Color(0, 0, 0, 0);
+            RuntimePreviewGenerator.OrthographicMode = true;
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs:
```diff
@@ -251,6 +252,22 @@ public void SetContainerUnchecked(Container newContainer)
+            Texture2D texture = RuntimePreviewGenerator.GenerateModelPreviewWithShader(this.transform, Shader.Find("Legacy Shaders/Diffuse"), null, 128, 128, false);
+
+            _sprite = Sprite.Create(texture, new Rect(0, 0, texture.width, texture.height), new Vector2(0.5f, 0.5f), 100);
+            
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Storage/UI/ItemDisplay.cs:
```diff
@@ -74,6 +74,9 @@ public void OnPointerClick(PointerEventData eventData)
         public void OnBeginDrag(PointerEventData eventData)
         {
+            // Only allow to drag with a left click.
+            if(eventData.button != PointerEventData.InputButton.Left) return;
```
add space after if

stilnat committed:
Fix the UI when many player are in game.

The UI is working correctly with multiple players, simply by not setting up the Inventory reference of the InventoryUI when the code is not executed by the owner.

stilnat committed:
correct a few formatting issues

stilnat commented:
The UI is now working on any player, client or host.

TaylorNAlbarnaz reviewed (approved):
The problem was properly fixed and it works as expected now
